java-squid/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -0,0 +1,377 @@
+  ConstraintManager constraintManager;
+
+  public static class MaximumStepsReachedException extends RuntimeException {
+    public MaximumStepsReachedException(String s) {
+      super(s);
+    }
+  }
+
+  public ExplodedGraphWalker(JavaFileScannerContext context) throws MaximumStepsReachedException {
+    alwaysTrueOrFalseChecker = new ConditionAlwaysTrueOrFalseChecker();
+    this.checkerDispatcher = new CheckerDispatcher(this, context, Lists.newArrayList(alwaysTrueOrFalseChecker, new NullDereferenceChecker()));
+  }
+
+  @Override
+  public void visitMethod(MethodTree tree) {
+    super.visitMethod(tree);
+    BlockTree body = tree.block();
+    if (body != null) {
+      execute(tree);
+    }
+  }
+
+  private void execute(MethodTree tree) {
+    checkerDispatcher.init();
+    CFG cfg = CFG.build(tree);
+    explodedGraph = new ExplodedGraph();
+    constraintManager = new ConstraintManager();
+    workList = new LinkedList<>();
+    LOG.debug("Exploring Exploded Graph for method " + tree.simpleName().name() + " at line " + ((JavaTree) tree).getLine());
+    programState = ProgramState.EMPTY_STATE;
+    Iterable<ProgramState> startingStates = Lists.newArrayList(programState);
+    for (final VariableTree variableTree : tree.parameters()) {
+      // create
+      final SymbolicValue sv = constraintManager.eval(programState, variableTree);
+      startingStates = Iterables.transform(startingStates, new Function<ProgramState, ProgramState>() {
+        @Override
+        public ProgramState apply(ProgramState input) {
+          return put(input, variableTree.symbol(), sv);
+        }
+      });
+
+      if (variableTree.symbol().metadata().isAnnotatedWith("javax.annotation.CheckForNull")) {
+        startingStates = Iterables.concat(Iterables.transform(startingStates, new Function<ProgramState, List<ProgramState>>() {
+          @Override
+          public List<ProgramState> apply(ProgramState input) {
+            return Lists.newArrayList(
+              ConstraintManager.setConstraint(input, sv, ConstraintManager.NullConstraint.NULL),
+              ConstraintManager.setConstraint(input, sv, ConstraintManager.NullConstraint.NOT_NULL));
+          }
+        }));
+
+      }
+    }
+    for (ProgramState startingState : startingStates) {
+      enqueue(new ExplodedGraph.ProgramPoint(cfg.entry(), 0), startingState);
+    }
+    steps = 0;
+    while (!workList.isEmpty()) {
+      steps++;
+      if (steps > MAX_STEPS) {
+        throw new MaximumStepsReachedException("reached limit of " + MAX_STEPS + " steps for method " + tree.simpleName().name() + "in class " + tree.symbol().owner().name());
+      }
+      // LIFO:
+      node = workList.removeFirst();
+      programPosition = node.programPoint;
+      if (/* last */programPosition.block.successors().isEmpty()) {
+        // not guaranteed that last block will be reached, e.g. "label: goto label;"
+        // TODO(Godin): notify clients before continuing with another position
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.internal.sonarsource.com/coding_rules#rule_key=squid%3AS1135)